### ISSUE TITLE
[HOLD for Payment 2023-09-20][$2000]  Can't parse deep link with email param.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16762]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Try to open user detail from deep link on browser (for example: https://staging.new.expensify.com/details?login=testing@gmail.com).
2. Now go to any report, paste that link and click send.
3. Notice that it can't parse the whole link, and when user try to click on it, it opens the email app with the wrong email filled.

## Expected Result: 
App should parse the deep link detail with email param.

## Actual Result: 
App can't parse the deep link detail with email param. The URL prefix (`https://`) is not part of the link, and we instead link incorrectly.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.92-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/228922658-ef1627e6-9f3e-4f91-a52a-8cb26a5ea47f.mov

<img width="952" alt="Untitled" src="https://user-images.githubusercontent.com/43996225/228922720-e9c6c11b-59bf-4d1f-88fa-eabff471a1fd.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @hungvu193 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1680191331185739

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013659e025d6ce0e4f</li>
        <li>Upwork Job ID: 1644482548992663552</li>
        <li>Last Price Increase: 2023-06-26</li>
<li>Automatic offers: </li>
<ul>
<li>0xmiroslav | Reviewer | 26524537</li>
<li>Antasel | Contributor | 26524539</li>
<li>hungvu193 | Reporter | 26524541</li>
</ul></li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sakluger (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@sakluger Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@sakluger Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@sakluger Still overdue 6 days?! Let's take care of this!

**Comment by sakluger** [comment]
I was able to reproduce this issue, feels like a bug worth solving!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~013659e025d6ce0e4f

**Comment by MelvinBot** [comment]
Current assignee @sakluger is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @0xmiroslav (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @johnmlee101 (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by alitoshmatov** [comment]
I couldn't reproduce this issue. I am not sure how exactly you are copying the url. Can you provide more details, or elaborate on copying step a little bit more

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Deep links which contain an email address incorrectly parsed by the composer resulting in a broken link in the chat.

### What is the root cause of that problem?
The ExpensiMark parser is processing the `autoEmail` rule before the `autoLink` rule, which results in a broken link, because the email part of the deep link is parsed first, hence the full link will never be recognised.
The currently defined order is like so.
https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L95-L130

### What changes do you think we should make in order to solve the problem?
Modify the code to move the `autoLink` rule to come before the 'autoEmail` rule. This will fix the problem illustrated in this issue. We can also modify the 'link' rule to precede the `email` rule in the lines prior.

### What alternative solutions did you explore? (Optional)
None.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App can't parse the deep link detail with email param. The URL prefix (https://) is not part of the link, and we instead link incorrectly.

### What is the root cause of that problem?
The root cause is here https://github.com/Expensify/expensify-common/blob/3590b7371fe8fa73288a6625daad985e7fcfb057/lib/ExpensiMark.js#L101 where we're parsing the `autoEmail` before the `autoLink`. 

So when we're trying to parse `https://staging.new.expensify.com/details?login=testing@gmail.com`, it will parse `//staging.new.expensify.com/details?login=testing@gmail.com` as an email. Then when we try to apply the `autolink`, it will not work because the `<a>` tags are already inserted by the `autoEmail` rule.

The ordering of `autoEmail` and `autolink` are like that for a reason, to avoid the auto link rule to apply to a part of the email address. 

Let's consider this email `google.com@gmail.com`, if the `autolink` rule is before the `autoEmail` rule, the `google.com` will become an auto link and it will not be parsed as an email.

### What changes do you think we should make in order to solve the problem?

We need to do 2 things:
1. Put the auto link rule before the autoEmail rule, this can be done simply by reodering this rule https://github.com/Expensify/expensify-common/blob/3590b7371fe8fa73288a6625daad985e7fcfb057/lib/ExpensiMark.js#L101 and this rule https://github.com/Expensify/expensify-common/blob/3590b7371fe8fa73288a6625daad985e7fcfb057/lib/ExpensiMark.js#L114

> The ordering of `autoEmail` and `autolink` are like that for a reason, to avoid the auto link rule to apply to a part of the email address. 

2. Address this issue. 
For this, we can modify the `URL_WEBSITE_REGEX` here https://github.com/Expensify/expensify-common/blob/f37b3bc00d1275f232c7ddb8bbcb5b6f69310918/lib/Url.js#L3 such that it will not match if the domain is followed by an email signature (a.k.a `@domain.tld`). This can be done by updating this rule https://github.com/Expensify/expensify-common/blob/f37b3bc00d1275f232c7ddb8bbcb5b6f69310918/lib/Url.js#L3
to 
```
const URL_WEBSITE_REGEX = `(https?:\\/\\/)?((?:www\\.)?([-a-z0-9])+?\\.)+(?:${TLD_REGEX})(?!@[a-zA-Z0-9-]+?(\.[a-zA-Z]+)+)(?:\\:\\d{2,4}|\\b|(?=_))`;
```
(note the `(?!@[a-zA-Z0-9-]+?(\.[a-zA-Z]+)+)` to match the `@domain.tld` signature, it's the same regex we're using to match that part of an `autoEmail`)

### What alternative solutions did you explore? (Optional)
- We can improve the regex in `2` if there're special edge cases that need to be addressed (special characters, ...), for example by adding `[a-zA-Z0-9.!$%&'*+=?^_{|}~-]*` to match the email-compatible part between the matched domain and the `@domain.tld` signature, but the main point is the same. If an accidental URL that belongs to an email address is not parsed as an auto link, it will be safe to reorder the rules.

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
if we parse link with email param, it parased as email not link.
### What is the root cause of that problem?
the REGEXP of `autoEmail` allow to parse email in link text start with http or https.
### What changes do you think we should make in order to solve the problem?
edit REGEXP of `autoEmail` to prevent match emails in any text start with http or https.
add negative look behind `(?<!http(s?):\S*)`  
```javascript
// replace this
(?![^<]*>|[^<>]*<\\/)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))

// with this
(?![^<]*>|[^<>]*<\\/)(?<!http(s?):\S*)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))
```

### What alternative solutions did you explore? (Optional)

**Comment by anaskhraza** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

App can't parse the deep link detail with email param. The URL prefix (https://) is not part of the link, and we instead link incorrectly.

### What is the root cause of that problem?

1 - When the user tries to send a deeplink with email as param, the parser rule that we have defined in the library [rule autoEmail](https://github.com/Expensify/expensify-common/blob/3590b7371fe8fa73288a6625daad985e7fcfb057/lib/ExpensiMark.js#L95-L107) should automatically links the email. However our current rule can only work with HTML tags not the raw URL string.

2 - The placement of autoemail rule is also contributing to the problem, as it's been placed before autolink.

### What changes do you think we should make in order to solve the problem?

Step 1 - Update the regex rule to prevent match emails if string contains http or https

Step 2 - Reorder autoEmail rule to check if the matching string is a URL, we will use URL_REGEX and don't need to format to an email link

### What alternative solutions did you explore? (Optional)
none

**Comment by MelvinBot** [comment]
📣 @anaskhraza! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by anaskhraza** [comment]
Contributor details
Your Expensify account email: m.khurshid1991@outlook.com
Upwork Profile Link: https://www.upwork.com/freelancers/~013888b99767abfe1e

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by MelvinBot** [comment]
@johnmlee101, @sakluger, @0xmiroslav Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by sakluger** [comment]
Hey @0xmiroslav @johnmlee101 how do the proposals look so far?

**Comment by MelvinBot** [comment]
@johnmlee101 @sakluger @0xmiroslav this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by johnmlee101** [comment]
@0xmiroslav Do you mind reviewing the proposals please? 🙇

**Comment by MelvinBot** [comment]
@johnmlee101 @sakluger @0xmiroslav this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by johnmlee101** [comment]
@0xmiroslav Whats the latest here?

**Comment by sakluger** [comment]
@0xmiroslav can you please give us an update on the progress of your review? If there are issues with any of the proposals, it would help to know so that we could get updated or new proposals.

**Comment by akinwale** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1500744733)

@0xmiroslav Please have a look at my updated proposal. Thanks.

**Comment by ahmedGaber93** [comment]
@0xmiroslav 
autoEmail ignoring works fine for this url
`https://www.expensify.com?name=test&email=test@expensify.com`
try it [here](https://regex101.com/r/7ZiGq8/1)

**Comment by akinwale** [comment]
> @akinwale I still feel it's workaround, case-by-case fix. Your updated solution will fail this case:
> 
> ```
> [test ]](test@expensify.com)
> ```
> 
> Email should be parsed.

I missed out word boundaries in my update. I have updated my proposal which should handle this case properly. The regex should now cover all scenarios properly.

Is there a list of standard test cases that we can test against?

**Comment by akinwale** [comment]
@0xmiroslav

Here's the updated regex.
```
`(?![^<]*>|[^<>]*<\\/)(^| |\\b)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))`
```

Test scenarios. All testable as individual lines or as a single comment.
```
test@expensify.com https://www.expensify.com
bruh test@expensify.com https://www.expensify.com
https://staging.new.expensify.com/details?login= testing@gmail.com bacon antoher@gmail.com
multiple spaces https://www.expensify.com    test@expensify.com
[test ]](test@expensify.com)
[testing](test@expensify.com)
another scenario [test@expensify.com]
checking boundaries \test@expensify.com
https://staging.new.expensify.com/details?name=test&email= testing@gmail.com
final@expensify.com last check
```

**Comment by akinwale** [comment]
> @akinwale `[test ]](test@expensify.com)` still doesn't work for me <img alt="Screenshot 2023-04-26 at 12 29 13 PM" width="72" src="https://user-images.githubusercontent.com/97473779/234548714-22bbf49f-3b2e-4418-b9ec-671230ffa3c8.png">
> 
> Proposal doesn't need to be instant. If you're not confident, please test enough for a few days and then submit.

That's odd. My regex replacement does not remove the email match at all. Please note that you have to make use of the `replacement` method as shown in my proposal, in addition to the new regex.

So you should have the following before testing:
```
{
    name: 'autoEmail',
    regex: new RegExp(
        `(?![^<]*>|[^<>]*<\\/)(^| |\\b)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))`,
        'gim',
    ),
    replacement: (match, g1, g2) => {
        let rep = `<a href="mailto:${g2}">${g2}</a>`;
        if (match.startsWith(' ')) {
            rep = ' ' + rep;
        }
        return rep;
    }
}
```

I have recorded a video showing a demo.

[16762-demo.webm](https://user-images.githubusercontent.com/4160319/234549971-aae054a5-1945-4a48-aab4-5291308c4c2d.webm)

**Comment by akinwale** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1500744733)

@0xmiroslav Updated proposal with the final rule and test cases.

Please note that in addition to the `regex`, the `replacement `method has also changed, handling 3 capture groups instead of 2.

**Comment by MelvinBot** [comment]
Current assignee @0xmiroslav is eligible for the Internal assigner, not assigning anyone new.

**Comment by johnmlee101** [comment]
I think we're close to a solution, @0xmiroslav what do you think of the latest updated proposals?

**Comment by sakluger** [comment]
@johnmlee101 do you think we're close enough to switch back to external?

**Comment by akinwale** [comment]
> I reviewed recent 2 proposals after last review. Still doesn't work for some cases. @tienifr @akinwale please run automated tests (`npm run test`) in expensify-common repo. You will easily find regression cases. Please submit updated proposal only when tests pass.

@0xmiroslav Thanks for the update. I was initially looking for these tests, so that I could run them, as I was testing directly in the browser. I have updated the regex in my proposal, and all test cases pass now. I also like the idea of bumping the price. 😅 

Here are the results.

![16762-npm-run-test](https://user-images.githubusercontent.com/4160319/235040591-7bcb635c-9195-40c9-a186-c8ab11eaeb48.png)

### Demo video

[16762-final-demo.webm](https://user-images.githubusercontent.com/4160319/235040684-23b5b9ce-e85a-4d8b-a886-55402bddf9af.webm)

**Comment by ahmedGaber93** [comment]
I update my [proposal](https://github.com/Expensify/App/issues/16762#issuecomment-1501014993), and it works fine and pass all test.
But I recently find negative look behind (?<!) supported in ECMAScript 9(2018) which we use ES6 as I know (I think this can help others to stop using it)

**Comment by ahmedGaber93** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1501014993)


edit REGEXP of `autoEmail` to be match **optional** `http\\S*=`  and use it in replacement case like this.
```javascript
{
    name: 'autoEmail',
    regex: new RegExp(
        `(?![^<]*>|[^<>]*<\\/)((?:http\\S*=)?)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))`,
        'gim',
    ),
    replacement: (match, g1, g2) => {
        return g1 ? match : `<a href="mailto:${g2}">${g2}</a>`;
    },
},
```

<details>
<summary>test text</summary>

```
https://www.expensify.com?name=test&email=testing@gmail.com
https://www.expensify.com?name=test&email= testing@gmail.com
test@expensify.com https://www.expensify.com
bruh test@expensify.com https://www.expensify.com
https://www.expensify.com test@expensify.com
https://staging.new.expensify.com/details?login= testing@gmail.com bacon antoher@gmail.com
multiple spaces https://www.expensify.com    test@expensify.com
[test ]](test@expensify.com)
another scenario [test@expensify.com]
checking boundaries \test@expensify.com
https://staging.new.expensify.com/details?name=test&email= testing@gmail.com
final@expensify.com last check
testing@gmail.com google.com@gmail.com tests fine
https://www.expensify.com?name=test&email=testing@gmail.com
random@gmail.com https://www.expensify.com?name=test&email=testing@gmail.com random2@gmail.com

```
</details>

all test passed in [expensify-common](https://github.com/Expensify/expensify-common) by npm run test

result 

https://user-images.githubusercontent.com/41129870/235198883-c1843f0b-8f7a-4fca-9f52-be6b3b27d6fd.mov

**Comment by johnmlee101** [comment]
Yeah I think I'm fine with bumping the price, it's a fairly complex, and the discussions have been active so I don't think we'll move this to internal just yet

**Comment by MelvinBot** [comment]
Current assignee @sakluger is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @0xmiroslav is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @johnmlee101 is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by sakluger** [comment]
> @sakluger as this is really not an easy issue, we can bump the price so that contributors investigate more time to avoid any regressions. or @johnmlee101 will work on this internally?

I agree, I've raised to $2000.

**Comment by sakluger** [comment]
@0xmiroslav any thoughts on the latest proposals?

**Comment by tienifr** [comment]
> I reviewed recent 2 proposals after last review. Still doesn't work for some cases.
@tienifr @akinwale please run automated tests (npm run test) in expensify-common repo. You will easily find regression cases.
Please submit updated proposal only when tests pass.

@0xmiroslav the tests failed due to new changes to the `URL_WEBSITE_REGEX` that causes the example regex I provided to be incorrect, not due to the solution itself.

[Proposal updated](https://github.com/Expensify/App/issues/16762#issuecomment-1500807823) to address that, all tests passing now.

<img width="565" alt="Screenshot 2023-05-02 at 21 15 28" src="https://user-images.githubusercontent.com/113963320/235693275-870f4d05-ecf4-4ce6-9cf7-e05055af2f24.png">
Thanks!

**Comment by sakluger** [comment]
@0xmiroslav could you please review the lastest proposals?

**Comment by sakluger** [comment]
All good, thanks for the context. That's fine since EC3 issues take priority right now.

**Comment by sakluger** [comment]
EC3 is still top priority.

**Comment by johnmlee101** [comment]
It's been a week. Are you still reviewing proposals?

**Comment by sakluger** [comment]
EC3 is happening! And deploys are paused for the rest of the week. @0xmiroslav hopefully that means there will be some time to review these proposals.

**Comment by sakluger** [comment]
Hey @0xmiroslav, can you please prioritize reviewing these proposals? The latest proposal updates have been waiting 3 weeks, so we need to start moving this forward.

**Comment by johnmlee101** [comment]
Bump! @0xmiroslav

**Comment by akinwale** [comment]
> @akinwale your solution failing test cases. please update proposal based on latest codebase. To confirm this is your regex, right?
> 
> ```
> (?![^<]*>|[^<>]*<\\/)(^| |\\b)${CONST.REG_EXP.MARKDOWN_EMAIL}(?![^<]*(<\\/pre>|<\\/code>|<\\/a>))
> ```

No.

Please check my [proposal](https://github.com/Expensify/App/issues/16762#issuecomment-1500744733) for my most up to date regex and replacement. It was updated over a month ago. Thanks.

**Comment by sakluger** [comment]
Bump @0xmiroslav. I noticed that you have a lot of issues assigned currently - let me know if you'd like me to reassign, it's no problem.

**Comment by sakluger** [comment]
Readded BZ label since I'm OOO next week.

**Comment by flaviadefaria** [comment]
@0xmiroslav do you think you can this assigned to a contributor by tomorrow?

**Comment by flaviadefaria** [comment]
@0xmiroslav can you give an update?

**Comment by flaviadefaria** [comment]
@0xmiroslav it's been more than 10 days here without hearing from you. What's going on?
Also @sakluger as you're back I'm returning this to you.

**Comment by crystal6117** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App is not parsing urls that include email addresses as a param in itself.

### What is the root cause of that problem?
The reason is that the given url is recognized as an email address not a deep link since the `autoEmail` rule is executed early rather than `autoLink` in the ExpensiMark parser. 

### What changes do you think we should make in order to solve the problem?
Simply, we can solve this problem by placing the `autoLink` rule before the `autoEmail` one. 

Additionally, we need to update the current email regex, so that it does not allow email addresses that contain the forward slash ("/") character. According to the RFC 5322 specification, the forward slash ("/") character is not allowed in the local part of an email address.

### What alternative solutions did you explore? (Optional)
None

**Comment by crystal6117** [comment]
Contributor details
Your Expensify account email: vividsoftdev@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01c86b3fc66ec7f35f

**Comment by crystal6117** [comment]
Hello, @0xmiroslav 

Here is my updated solution for this issue.
- First need to update email regular expression to avoid "/" and "=" characters. Like this
```MARKDOWN_EMAIL: "([a-zA-Z0-9.!#$%&'+?^`{|}-][a-zA-Z0-9.!#$%&'*+?^_`{|}~-]*@[a-zA-Z0-9-]+?(\\.[a-zA-Z]+)+)"```
- Then need to update `rules` in https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js
   First put `autoLink rule before `autoEmail` one. Then divide `autoLink` rule into 2 another rules for `LOOSE_URL_REGEX` and `URL_REGEX` respectively to give higher priority to `LOOSE_URL_REGEX` rule than `URL_REGEX'.
   `autoLink` -> `autoLink` for LOOSE_URL_REGEX and `autoLink2` for URL_REGEX

P.S: I have some suggestion for current test case.
 1. The expected result for ```staging.new.expensify.com/details?login=testing@gmail.com ``` is not correct in `Urls with query params` Clause.
Current result: 
```
<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com</a>
```
However my suggestion is 
```
<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a>
```
2. I don't think ```<a href="mailto:email=test@expensify.com">email=test@expensify.com</a>``` is not correct.
Here is a result for it in slack.
![Screenshot 2023-06-21 at 2 39 27 PM](https://github.com/Expensify/App/assets/55975468/5a7f0651-a00f-4c70-bd71-8fde86b0d67b)

I confirmed this solution pass all the test.

Thank you.
Hristijan

**Comment by crystal6117** [comment]
So how about my solution?

**Comment by crystal6117** [comment]
https://github.com/Expensify/App/issues/16762#issuecomment-1601568281

**Comment by crystal6117** [comment]
Oh, sorry. Wrong link. Here it is. https://github.com/Expensify/App/issues/16762#issuecomment-1598476985
And if you want, I can submit a new proposal with the template for my updated solution.

**Comment by crystal6117** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1601568281)

### Please re-state the problem that we are trying to solve in this issue.
App is not parsing urls that include email addresses as a param in itself.

### What is the root cause of that problem?
There were some small issues in a regular expression for email and the order to parse various string was not correct

### What changes do you think we should make in order to solve the problem?
- First need to update email regular expression to avoid "/" and "=" characters. Like this
```MARKDOWN_EMAIL: "([a-zA-Z0-9.!#$%&'+?^`{|}-][a-zA-Z0-9.!#$%&'*+?^_`{|}~-]*@[a-zA-Z0-9-]+?(\\.[a-zA-Z]+)+)"```
- Then need to update `rules` in https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js
   First put `autoLink` rule before `autoEmail` one. Then divide `autoLink` rule into 2 another rules for `LOOSE_URL_REGEX` and `URL_REGEX` respectively to give higher priority to `LOOSE_URL_REGEX` rule than `URL_REGEX`.
   `autoLink` -> `autoLink1` for LOOSE_URL_REGEX and `autoLink2` for URL_REGEX

### What alternative solutions did you explore? (Optional)
None

**Comment by crystal6117** [comment]
Hello, @0xmiroslav. Will you give me reviews for my proposal?

**Comment by crystal6117** [comment]
Can't you accept [this proposal](https://github.com/Expensify/App/issues/16762#issuecomment-1601613643)?

**Comment by crystal6117** [comment]
Just updated my proposal with detailed contents.

**Comment by crystal6117** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1601613643)

### Please re-state the problem that we are trying to solve in this issue.
App is not parsing urls that include email addresses as a param in itself.

### What is the root cause of that problem?
Firstly, the current regular expression to parse email is allowing "=" and "/" characters. By the way, the characters make an important role in a deep link, so the regular expression should not allow the ones.
Secondly, the order of rules is wrong. The `autoLink` rule should be put before the `autoEmail` one because a link can contain email, but email can’t contain a link.

### What changes do you think we should make in order to solve the problem?
According to the root cause of this problem, need to make 2 changes to solve this one primarily.

Firstly need to update the current regular expression. Here is updated one.
```
([a-zA-Z0-9.!#$%&'+?^`{|}-][a-zA-Z0-9.!#$%&'*+?^_`{|}~-]*@[a-zA-Z0-9-]+?(\\.[a-zA-Z]+)+)
```

Secondly, need to put the `autoLink` rule before the `autoEmail` one in the [rule list](https://github.com/Expensify/expensify-common/blob/3590b7371fe8fa73288a6625daad985e7fcfb057/lib/ExpensiMark.js#L16). 

Thirdly, the `autoLink` rule should be divided into 2 rules. 
The regular expression engine does not consider overlapping matches. In this case, when "example.com/https://www.expensify.com" is matched, the engine moves forward in the string and does not consider the "https://www.expensify.com" part as a separate match. Therefore, only the longer match, "example.com/https://www.expensify.com", is returned. Here, `MARKDOWN_URL_REGEX` rule is `(${LOOSE_URL_REGEX}|${URL_REGEX})`, and "example.com/https://www.expensify.com" is parsed from `URL_REGEX`
 and "https://www.expensify.com" is parsed from `LOOSE_URL_REGEX`.
So my solution is to divide the `autoLink` rule into 2 rules for `LOOSE_URL_REGEX` and `URL_REGEX` respectively. When the rule is dividing into others, the rule for `LOOSE_URL_REGEX` should be put before the another one for `URL_REGEX` to give higher priority to the one for `LOOSE_URL_REGEX`.

P.S. I confirmed that all test cases were passed by applying this solution.

### What alternative solutions did you explore? (Optional)
none.

**Comment by sakluger** [comment]
If we haven't accepted a proposal by the end of the week (so end of the day tomorrow), I'm going to reach out to our expert contributors for help.

**Comment by ahmedGaber93** [comment]
@0xmiroslav I think you missed whitespace in test cases `Bad urls > urlTestReplacedString`

```
+ '/details?login=<a href="mailto:testing@gmail.com">testing@gmail.com</a>'
+ '/details?login=<a href="mailto:testing@gmail.com">testing@gmail.com</a> ' //<<<<< here
```

**Comment by ahmedGaber93** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1501014993)

**Comment by BeeMargarida** [comment]
Hey :wave: I’m Ana from Callstack, an expert agency, and I would like to help close this issue out. I see that there are a lot of proposals and an additional test suite. I'll go through the proposals and test with the existing and @0xmiroslav test suites and try make a summary of the pros/cons of each proposal. Does it sound good?

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Can't parse deep link with email param.

### What is the root cause of that problem?

The root cause of this issue is that we apply [autoEmail](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L127) rule first and then apply [autoLink](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L141) rule to parse most use cases of `autoEmail` syntax and `autoLink` syntax correctly. The order of the two rule is subtle because

1. An `autoEmail` syntax includes a substring of `autoLink` syntax.
2. The `autoLink` rule relies on the html anchor tag of `autoEmail` (after parsing) to avoid parsing the domain of email, see negative lookahead group of  [autoLink](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L141) which uses `</a>`.

But the two rules fail to parse an `autoLink` syntax contains email in the param part because the `autoEmail` rule is applied first without checking if it's part of an `autoLink` syntax.

### What changes do you think we should make in order to solve the problem?

Note that the current [autoLink](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L141) can parse a link with email param, like

```
https://staging.new.expensify.com/details/test@expensify.com
```

if it's applied before the [autoEmail](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L127) rule. But we can't simply move the [autoLink](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L141) rule before the [autoEmail](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L127) rule for the reason we point out int the RCA part.

To fix this issue, we can add an `autolinkWithEmailParam` rule before the [autoEmail](https://github.com/Expensify/expensify-common/blob/f9985d572c772757885ff59ea5c82b1050c6b3f7/lib/ExpensiMark.js#L127) rule to parse `autoLink` syntax includes email param only. The new rule will be like

```js
// The only difference between this rule and 'autoLink' rule is that this rule only parses link includes email param
{
    name: 'autolinkWithEmailParam',

    process: (textToProcess, replacement) => {
        const regex = new RegExp(
            `(?![^<]*>|[^<>]*<\\/)([_*~]*?)${MARKDOWN_URL_REGEX}\\1(?!((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`,
            'gi',
        );
        return this.modifyTextForUrlLinks(regex, textToProcess, replacement);
    },

    replacement: (match, g1, g2) => {
      
        // We want to only parse link includes email param
        if (!Str.extractEmail(g2)) {
            return `${g1}${g2}${g1}`;
        }
        const href = Str.sanitizeURL(g2);
        return `${g1}<a href="${href}" target="_blank" rel="noreferrer noopener">${g2}</a>${g1}`;
    },
},
```

This solution parses most test cases provided by https://github.com/Expensify/App/issues/16762#issuecomment-1598810396 except for following two cases of `autoEmail` syntax 

```
+ '/details?login=testing@gmail.com '
+ '?name=test&email=test@expensify.com';
```

<img width="600" alt="image" src="https://github.com/Expensify/App/assets/117511920/91689402-b689-4045-9390-33a67d847ca9">

I think the above two test cases are arguable because it seems, from RFC, special characters like

```
!#$%&'*+-/=?^_`{|}~
```

are all legal in the local part of an email, see this [SO](https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address) discussion.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by eh2077** [comment]
@0xmiroslav I added a new solution which passed most test cases you provided except for the following two arguable test cases of `autoEmail`
```
+ '/details?login=testing@gmail.com '
+ '?name=test&email=test@expensify.com';
```
From the discussion of this [SO](https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address), I learned that special characters like

```
!#$%&'*+-/=?^_`{|}~
```
are all legal in the local part of an email.

Looking forward to your feedback on this.

**Comment by ahmedGaber93** [comment]
Add some comments and explanations to make the proposal easy to understand [My proposal](https://github.com/Expensify/App/issues/16762#issuecomment-1501014993).

**Comment by BeeMargarida** [comment]
Update: Still investigating and testing all proposals, will post a summary soon

**Comment by BeeMargarida** [comment]
Update: tTook a while to gather all the info and screenshots and make all the summaries. Since I've exceeded my working time, I'll post this tomorrow morning

**Comment by BeeMargarida** [comment]
## Summary of all proposals

There is one question that should be answered and that will change the validity of some proposals:

> Should we be following the RFC specification of a valid email?

Following this [SO](https://stackoverflow.com/a/2049510) post shared by @eh2077 and the following [RFC link 1](https://www.rfc-editor.org/rfc/rfc2822#page-17) and [RFC link 2](https://www.rfc-editor.org/rfc/rfc2822#page-12), there are certain incomplete URLs that are actually valid emails. Example:
    - `/details?login=testing@gmail.com`
    - `//staging.new.expensify.com/details?login=testing@gmail.com`
    - `//www.expensify.com?name=test&email=test@expensify.com`
    - `https//www.expensify.com@gmail.com` (check that it does not contain `:`)

If we follow this specification, some of the additional test cases for `Bad urls` are wrong and the one - `staging.new.expensify.com/details?login=testing@gmail.com` - for `Urls with query params` is also wrong.

Note: Use [this Scriptular](https://scriptular.com/#%5E(%3F!%5C.)(%22%22(%5B%5E%22%22%5Cr%5C%5C%5D%7C%5C%5C%5B%22%22%5Cr%5C%5C%5D)*%22%22%7C(%5B-a-z0-9!%23%24%25%26'*%2B%2F%3D%3F%5E_%60%7B%7C%7D~%5D%7C(%3F%3C!%5C.)%5C.)*)(%3F%3C!%5C.)%40%5Ba-z0-9%5D%5B%5Cw%5C.-%5D*%5Ba-z0-9%5D%5C.%5Ba-z%5D%5Ba-z%5C.%5D*%5Ba-z%5D%24%7C%7C%7C%7Cgim%7C%7C%7C%7C%5B%22%2F%2Fwww.expensify.com%3Fname%3Dtest%26email%3Dtest%40expensify.com%22%2C%22%22%5D) to test the validity of an email that follows the RFC specification.

One point that I find odd is that the `REG_EXP.MARKDOWN_EMAIL` is correctly following the RFC, but the `EMAIL_BASE_REGEX` does not. Is there a reason for this?

### Proposals

#### By @akinwale

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1500744733

Passes the test suite (including the extra ones by @0xmiroslav) :x:
    - If we completely follow the specification (and remove the tests that do not follow it), it passes the tests

Problems (if we don't follow the specification):
- does not support emails as query params in incomplete URLs, like `staging.new.expensify.com/details?login=testing@gmail.com`
- bad URL

#### By @hoangzinh

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1500804673

Passes the test suite (including the extra ones by @0xmiroslav) :x:

Problems (if we don't follow the specification):
- does not support emails as a path params, like `https://staging.new.expensify.com/details/test@expensify.com`
- (if we don't follow the specification) bad URLs test suite

#### By @tienifr

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1500807823

Passes the test suite (including the extra ones by @0xmiroslav) :x:

Problems:
- does not support email followed by a link, like `test@expensify.com/https://www.expensify.com`, it converts the email part correctly but it then does not convert the link
  - This is due to the change related to the `URL_WEBSITE_REGEX`, which is ignoring everything that has an email domain after
- (if we don't follow the specification) it fails on the last 2 examples of Bad URL test case:
  - for example, for `/details?login?email@email.com`, it transforms into `<a href=\"mailto:/details?login=email@gmail.com\">/details?login=email@gmail.com</a>`

#### By @ahmedGaber93

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1501014993

Passes the test suite (including the extra ones by @0xmiroslav) ✅ 

The simpler solution (less code changes). It's similar in theory to @hoangzinh proposal but it works for all test cases.


#### By @anaskhraza

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1501099598

Can't be tested since no code was provided.

Passes the test suite (including the extra ones by @0xmiroslav) :x:

Problems:
--

#### By @crystal6117

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1601613643

Explanation: https://github.com/Expensify/App/issues/16762#issuecomment-1603237525

Passes the test suite (including the extra ones by @0xmiroslav) ✅ 

Notes:
- The autolink split is hard to understand, but it does fixes the issue of the example you mentioned: `example.com/https://www.expensify.com`, although it seems like that is another problem in itself. Although it's confusing without a space between them (see GIF), perhaps the Github approach of not replacing the first address would be better.

https://github.com/Expensify/App/assets/25725586/2bf87107-3c12-4fe0-a75e-fba0e5a949a1

#### By @eh2077

Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1605250961

Passes the test suite (including the extra ones by @0xmiroslav) :x:
  - If we completely follow the specification (and remove the tests that do not follow it), it passes the tests

Note:
- Even though it is mentioned that the last 2 test cases can be considered emails, there are more cases (like `//www.expensify.com?name=test&email=test@expensify.com` or `staging.new.expensify.com/details?login=testing@gmail.com`) that are also valid emails and not links. Since it follows the specification, it should not use `Str.extractEmail`, which uses a different regex for extracting the email.

### Conclusion

**If we follow the RFC specification**
There are 2 valid proposals: @akinwale and @eh2077

**If we ignore the RFC specification**
There are 2 valid proposals: @ahmedGaber93 and @crystal6117

**My opinion**
I think we should strive for RFC compliance, however, after some experimentation in other apps (Slack, Telegram, Gmail), all of them are not really RFC compliant in their text processing.

There's an issue to be discussed, which is the conversion of URL like strings to URL when they are also valid emails, for example: `//staging.new.expensify.com/details?login=testing@gmail.com`.

If we commit to not be RFC compliant, I recommend @crystal6117 solution since it's more complete and covers more edge cases like the last examples from the code snippet below. However, there are some changes in behaviour that must be clarified, since it's not clear if they were an issue before (see Note section in @crystal6117 proposal summary).

If we commit to be RFC compliant, there's not really a working proposal since the current logic is converting `//staging.new.expensify.com/details?login=testing@gmail.com` to an URL when in itself is a valid email. What's the priority? If we ignore this case, @eh2077 seems like the better solution since it covers more edge cases as well (URLs inside URLs).

```
test@expensify.com https://www.expensify.com
test@expensify.com-https://www.expensify.com
test@expensify.com/https://www.expensify.com
test@expensify.com?https://www.expensify.com
test@expensify.com>https://www.expensify.com
https://staging.new.expensify.com/details/test@expensify.com
staging.new.expensify.com/details

https://www.expensify.com?name=test&email=test@expensify.com
https://staging.new.expensify.com/details?login=testing@gmail.com
staging.new.expensify.com/details?login=testing@gmail.com
http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled

-https://www.expensify.com /https://www.expensify.com @https://www.expensify.com
expensify.com -expensify.com @expensify.com
https//www.expensify.com
//www.expensify.com?name=test&email=test@expensify.com
//staging.new.expensify.com/details?login=testing@gmail.com
/details?login=testing@gmail.com
?name=test&email=test@expensify.com

example.com/https://www.expensify.com

test@gmail.com staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com
test@gmail.com //staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com
test@gmail.com-https://staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com
```

<details>

<summary>Comparison between the two best solutions</summary>

| @akinwale | @eh2077  | @ahmedGaber93 | @crystal6117 |
| ------------ | ------------- | ------------ | ------------- |
|<img width="795" alt="akinwale" src="https://github.com/Expensify/App/assets/25725586/fc5c6687-9cc6-41c1-ac1b-611a33cd605c">|               |<img width="795" alt="eh2077" src="https://github.com/Expensify/App/assets/25725586/481cd171-99b4-4440-9cff-7d68a5510811">
|<img width="795" alt="ahmed" src="https://github.com/Expensify/App/assets/25725586/dbccee59-2c6e-4725-a7fe-7615eb38241d">|<img width="795" alt="crystal" src="https://github.com/Expensify/App/assets/25725586/7f202243-0545-4f40-840c-c1ba7118db55">|

</details>

**Comment by crystal6117** [comment]
Hello, @BeeMargarida. 
While following RFCs is not mandatory in software engineering, incorporating them into your development process can provide valuable guidance, improve interoperability, enhance quality, and align your software with widely accepted standards and best practices.

If following RFC for the email specification, all email addresses can contain several severe characters such as "/", "=" and "?", but many Mail applications such as Slack do not follow it as you mentioned.

**Comment by BeeMargarida** [comment]
@YacineF0001 Could you provide the code so that it's possible to test locally and compare with the other proposals?

**Comment by BeeMargarida** [comment]
Haven't gotten around to test it yet, will do it tomorrow. One thing is that this proposal seems heavily dependent on https://github.com/Expensify/App/issues/17387, which means that it should be approved there to also be applied here.

**Comment by sakluger** [comment]
Ah, it went overdue because I never assigned @BeeMargarida. Sorry about that!

**Comment by BeeMargarida** [comment]
@YacineF0001 I tested your solution, overall it works fine but I found a case where it didn't parse well: `test@gmail.com/https://example.com/google@email.com?email=asd@email.com`. For example, using @crystal6117 solution, this is correctly parsed.
At this point, it's just edge cases (in most cases it works well), but in addition to this, it heavily depends on https://github.com/Expensify/App/issues/17387, so this would have to be on hold.

@0xmiroslav Are you planning on leaving this on hold until https://github.com/Expensify/App/issues/17387 is resolved? It's related and might impact this if the regex is changed.

**Comment by sakluger** [comment]
If we're putting this on hold, I'll update the title and priority.

**Comment by sakluger** [comment]
No update, still waiting on hold issue.

**Comment by sakluger** [comment]
Still on hold for https://github.com/Expensify/App/issues/17387.

**Comment by sakluger** [comment]
Still on hold

**Comment by sakluger** [comment]
Still on hold.

**Comment by Antasel** [comment]
# Proposal
### Please re-state the problem that we are trying to solve in this issue.
Can't parse deep link with email param.

### What is the root cause of that problem?

The root cause is that first autoEmail rule is applied (so converted to `<a>~</a>`). Because of the `</a>` tag, autoLink rule is skipped.
Another cause is that current MARKDOWN_EMAIL regex pattern is validating email with '=', '/', '?' which is related to url .
but (=,/,?) whould be removed in email regex pattern with #17387 . 

### What changes do you think we should make in order to solve the problem?
I assumed that #17387 is fixed.

First, We can simply move autoEmail rule after autoLink rule, then parse link with email param first. so current issue is fixed.

Second, first step causes issue that link concatenated with email which is not param in the link, is not parsed. like test@expensify.com/https://www.expensify.com

To fix it, 
in autolink rule's modifyTextForUrlLinks() function, if the matched string faces @ sign before it, instead of simple abort, it will retry to apply autolink rule to the string except for like subdomain string after @ sign.
```
if ((match.index !== 0) && (textToCheck[match.index - 1] === '@')) {
      const domainRegex = new RegExp('^(([a-z-0-9]+\\.)+[a-z]{2,})(\\S*)', 'i');
      let domainMatch = domainRegex.exec(match[2]);

      if ((domainMatch !== null) && (domainMatch[3] !== '')) {
          shouldRetryByAtSign = true;
          replacedText = replacedText.concat(domainMatch[1] + this.replace(domainMatch[3], {filterRules: ['autolink']}));
      } else {
          abort = true;
      }
  }

```

<details>
  <summary>modifyTextForUrlLinks function would be as following</summary>

```
modifyTextForUrlLinks(regex, textToCheck, replacement) {
        let match = regex.exec(textToCheck);
        let replacedText = '';
        let startIndex = 0;

        while (match !== null) {
            // we want to avoid matching email address domains
            let abort = false;
            let shouldRetryByAtSign = false;

            // we want to avoid matching ending ) unless it is a closing parenthesis for the URL
            if (textToCheck[(match.index + match[2].length) - 1] === ')' && !match[2].includes('(')) {
                match[0] = match[0].substr(0, match[0].length - 1);
                match[2] = match[2].substr(0, match[2].length - 1);
            }

            // remove extra ) parentheses
            let brace = 0;
            if (match[2][match[2].length - 1] === ')') {
                for (let i = 0; i < match[2].length; i++) {
                    if (match[2][i] === '(') {
                        brace++;
                    }
                    if (match[2][i] === ')') {
                        brace--;
                    }
                }
                if (brace) {
                    match[0] = match[0].substr(0, match[0].length + brace);
                    match[2] = match[2].substr(0, match[2].length + brace);
                }
            }
            replacedText = replacedText.concat(textToCheck.substr(startIndex, (match.index - startIndex)));

            if ((match.index !== 0) && (textToCheck[match.index - 1] === '@')) {
                const domainRegex = new RegExp('^(([a-z-0-9]+\\.)+[a-z]{2,})(\\S*)', 'i');
                let domainMatch = domainRegex.exec(match[2]);
                
                if ((domainMatch !== null) && (domainMatch[3] !== '')) {
                    shouldRetryByAtSign = true;
                    replacedText = replacedText.concat(domainMatch[1] + this.replace(domainMatch[3], {filterRules: ['autolink']}));
                } else {
                    abort = true;
                }
            }

            if (abort) {
                replacedText = replacedText.concat(textToCheck.substr(match.index, (match[0].length)));
            } else if (!shouldRetryByAtSign) {
                const urlRegex = new RegExp(`^${LOOSE_URL_REGEX}$|^${URL_REGEX}$`, 'i');

                // `match[1]` contains the text inside the [] of the markdown e.g. [example](https://example.com)
                // At the entry of function this.replace, text is already escaped due to the rules that precede the link
                // rule (eg, codeFence, inlineCodeBlock, email), so we don't need to escape the text again here.
                // If the text `match[1]` exactly matches a URL, we skip translating the filterRules
                // So that special characters such as ['_', '*', '~'] are preserved in the text.
                const linkText = urlRegex.test(match[1]) ? match[1] : this.replace(match[1], {
                    filterRules: ['bold', 'strikethrough', 'italic'],
                    shouldEscapeText: false,
                });
                replacedText = replacedText.concat(replacement(match[0], linkText, match[2]));
            }
            startIndex = match.index + (match[0].length);

            // Now we move to the next match that the js regex found in the text
            match = regex.exec(textToCheck);
        }
        if (startIndex < textToCheck.length) {
            replacedText = replacedText.concat(textToCheck.substr(startIndex));
        }

        return replacedText;
    }

```

</details>
All test cases which @BeeMargarida used for are passed with my approach.

<details>
  <summary>Test cases</summary>

```
test('Test a url with email autolinks correctly', () => {
    let testString = 'https://staging.new.expensify.com/details?login=testing@gmail.com';
    let resultString = '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com</a>';
    testString = 'test@expensify.com/https://www.expensify.com';
    resultString = '<a href="mailto:test@expensify.com">test@expensify.com</a>/<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a>';
    expect(parser.replace(testString)).toBe(resultString);
});
test('Test a url autolinks correctly', () => {
    let testString = 'test@expensify.com https://www.expensify.com\n' +
    'test@expensify.com-https://www.expensify.com\n' +
    'test@expensify.com/https://www.expensify.com\n' +
    'test@expensify.com?https://www.expensify.com\n' +
    'test@expensify.com>https://www.expensify.com\n' +
    'https://staging.new.expensify.com/details/test@expensify.com\n' +
    'staging.new.expensify.com/details\n\n' +
    'https://www.expensify.com?name=test&email=test@expensify.com\n' +
    'https://staging.new.expensify.com/details?login=testing@gmail.com\n' +
    'staging.new.expensify.com/details?login=testing@gmail.com\n' +
    'http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled\n' +
    '-https://www.expensify.com /https://www.expensify.com @https://www.expensify.com\n' +
    'expensify.com -expensify.com @expensify.com\n' +
    'https//www.expensify.com\n' +
    '//www.expensify.com?name=test&email=test@expensify.com\n' +
    '//staging.new.expensify.com/details?login=testing@gmail.com\n' +
    '/details?login=testing@gmail.com\n' +
    '?name=test&email=test@expensify.com\n\n' +
    'example.com/https://www.expensify.com\n' +
    'test@gmail.com staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com //staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com-https://staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com';
    let resultString = '<a href="mailto:test@expensify.com">test@expensify.com</a> <a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' + 
    '<a href="mailto:test@expensify.com">test@expensify.com</a>-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="mailto:test@expensify.com">test@expensify.com</a>/<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="mailto:test@expensify.com">test@expensify.com</a>?<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' + 
    '<a href="mailto:test@expensify.com">test@expensify.com</a>&gt;<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details/test@expensify.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details/test@expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details</a><br /><br />'+ 
    '<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com?name=test&amp;email=test@expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '<a href="http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled" target="_blank" rel="noreferrer noopener">http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled</a><br />' +
    '-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> /<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> @https://www.expensify.com<br />' +
    '<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> -<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> @expensify.com<br />' +
    'https//<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com</a><br />' +
    '//<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com?name=test&amp;email=test@expensify.com</a><br />' +
    '//<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '/details?login=<a href="mailto:testing@gmail.com">testing@gmail.com</a><br />' +
    '?name=test&amp;email=<a href="mailto:test@expensify.com">test@expensify.com</a><br /><br />' +
    '<a href="https://example.com/https://www.expensify.com" target="_blank" rel="noreferrer noopener">example.com/https://www.expensify.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a> <a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a> //<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a>-<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a>';
    expect(parser.replace(testString)).toBe(resultString);
});

```

</details>

**Comment by sakluger** [comment]
@Antasel - we generally don't accept new proposals after a contributor is assigned and the Help Wanted label is removed. I do see that you were working on the issue we're holding this one for, so you probably have good context here. Are you saying that you think your proposal will work better than @BeeMargarida's proposal?

**Comment by Antasel** [comment]
Here is not confirmed proposal yet. so I posted my solution, which is covering all test cases on the basis of https://github.com/Expensify/expensify-common/pull/566 raised by me.

and @BeeMargarida did review contributors' proposals, and evaluated which was mostly expected to fix this issue.
but I am afraid nobody's solution passes the auto test cases what i mentioned. 

could you please check all approaches with autotest cases ? it should not make other cases failed

**Comment by BeeMargarida** [comment]
I've reviewed all proposals and it was not clear what the expected behaviour was, some proposals were more complete than others. This issue is dependent on https://github.com/Expensify/App/issues/17387, which has not yet been merged, so I don't think it's still worth it to move on with this one, since the email regex will possibly be changed.

**Comment by sakluger** [comment]
Thanks @BeeMargarida. I think we should wait until the hold issue is complete. If the proposal by @Antasel works well after the hold issue is deployed, then we can consider it.

**Comment by sakluger** [comment]
Still on hold for https://github.com/Expensify/App/issues/17387

**Comment by Antasel** [comment]
@sakluger @0xmiroslav 
#17387 's PR has been merged

**Comment by eh2077** [comment]
### Proposal
[Updated](https://github.com/Expensify/App/issues/16762#issuecomment-1605250961) proposal to apply `autolink` rule again after the `autoEmail` rule, which is able to pass all test cases provided by https://github.com/Expensify/App/issues/16762#issuecomment-1598810396

**Comment by sakluger** [comment]
I retested this and it's working as expected now! 🎉 @Antasel are you able to reproduce the issue in the latest build? If so, could you please share a video here?

**Comment by Antasel** [comment]
@sakluger 
we have still deep link parsing issue on the latest build .


https://github.com/Expensify/App/assets/139204153/d7d90418-b32b-4eae-8180-1f0b1f0e49fe

**Comment by sakluger** [comment]
Okay, thanks for testing all those edge cases!

@0xmiroslav I'd like your advice on how to move forward. Generally, we don't accept proposals while an issue is on hold, and I'm not sure if you've even reviewed @Antasel's proposal yet, what would you like to do?

**Comment by sakluger** [comment]
Oops, didn't mean to close the issue, my bad!

**Comment by sakluger** [comment]
Discussed in Slack (https://expensify.slack.com/archives/C01GTK53T8Q/p1693938665443249) - we're going to wait for @BeeMargarida's updated proposal, then decide which direction to go.

**Comment by ronibd0** [comment]
I agree with @BeeMargarida 
@Antasel proposal is good

**Comment by eh2077** [comment]
@BeeMargarida I also updated my proposal https://github.com/Expensify/App/issues/16762#issuecomment-1698955763. Would you like to have a look at it?

**Comment by Antasel** [comment]
@eh2077 Your proposal is dupe of @YacineF0001 's https://github.com/Expensify/App/issues/16762#issuecomment-1610420283
@BeeMargarida has already evaluated it [here](https://github.com/Expensify/App/issues/16762#issuecomment-1613212507)

**Comment by Antasel** [comment]
@0xmiroslav The following test cases are covering all of what you mentioned

<details>
<summary>Test Cases</summary>

```
test('Test a url with email autolinks correctly', () => {
    let testString = 'https://staging.new.expensify.com/details?login=testing@gmail.com';
    let resultString = '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com</a>';
    testString = 'test@expensify.com/https://www.expensify.com';
    resultString = '<a href="mailto:test@expensify.com">test@expensify.com</a>/<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a>';
    expect(parser.replace(testString)).toBe(resultString);
});
test('Test a url autolinks correctly', () => {
    let testString = 'test@expensify.com https://www.expensify.com\n' +
    'test@expensify.com-https://www.expensify.com\n' +
    'test@expensify.com/https://www.expensify.com\n' +
    'test@expensify.com?https://www.expensify.com\n' +
    'test@expensify.com>https://www.expensify.com\n' +
    'https://staging.new.expensify.com/details/test@expensify.com\n' +
    'staging.new.expensify.com/details\n\n' +
    'https://www.expensify.com?name=test&email=test@expensify.com\n' +
    'https://staging.new.expensify.com/details?login=testing@gmail.com\n' +
    'staging.new.expensify.com/details?login=testing@gmail.com\n' +
    'http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled\n' +
    '-https://www.expensify.com /https://www.expensify.com @https://www.expensify.com\n' +
    'expensify.com -expensify.com @expensify.com\n' +
    'https//www.expensify.com\n' +
    '//www.expensify.com?name=test&email=test@expensify.com\n' +
    '//staging.new.expensify.com/details?login=testing@gmail.com\n' +
    '/details?login=testing@gmail.com\n' +
    '?name=test&email=test@expensify.com\n\n' +
    'example.com/https://www.expensify.com\n' +
    'test@gmail.com staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com //staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com-https://staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com';
    let resultString = '<a href="mailto:test@expensify.com">test@expensify.com</a> <a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' + 
    '<a href="mailto:test@expensify.com">test@expensify.com</a>-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="mailto:test@expensify.com">test@expensify.com</a>/<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="mailto:test@expensify.com">test@expensify.com</a>?<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' + 
    '<a href="mailto:test@expensify.com">test@expensify.com</a>&gt;<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details/test@expensify.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details/test@expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details</a><br /><br />'+ 
    '<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com?name=test&amp;email=test@expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '<a href="http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled" target="_blank" rel="noreferrer noopener">http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled</a><br />' +
    '-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> /<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> @https://www.expensify.com<br />' +
    '<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> -<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> @expensify.com<br />' +
    'https//<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com</a><br />' +
    '//<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com?name=test&amp;email=test@expensify.com</a><br />' +
    '//<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '/details?login=<a href="mailto:testing@gmail.com">testing@gmail.com</a><br />' +
    '?name=test&amp;email=<a href="mailto:test@expensify.com">test@expensify.com</a><br /><br />' +
    '<a href="https://example.com/https://www.expensify.com" target="_blank" rel="noreferrer noopener">example.com/https://www.expensify.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a> <a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a> //<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a>-<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a>';
    expect(parser.replace(testString)).toBe(resultString);

    testString = 'test@gmail.com/https://example.com/google@email.com?email=asd@email.com';
    resultString = '<a href="mailto:test@gmail.com\">test@gmail.com</a>/<a href="https://example.com/google@email.com?email=asd@email.com" target="_blank" rel="noreferrer noopener">https://example.com/google@email.com?email=asd@email.com</a>';
    expect(parser.replace(testString)).toBe(resultString);

    testString = 'test@gmail.com/test@gmail.com/https://example.com/google@email.com?email=asd@email.com';
    resultString = '<a href="mailto:test@gmail.com">test@gmail.com</a>/<a href="mailto:test@gmail.com">test@gmail.com</a>/<a href="https://example.com/google@email.com?email=asd@email.com" target="_blank" rel="noreferrer noopener">https://example.com/google@email.com?email=asd@email.com</a>';
    expect(parser.replace(testString)).toBe(resultString);
});

```

</details>

**Comment by Antasel** [comment]
@0xmiroslav @sakluger 
The PR https://github.com/Expensify/expensify-common/pull/573 for expensify-common is ready for review

**Comment by Antasel** [comment]
@johnmlee101 
Could you help me to get expensify open-source channel invitation on Slack ?
I've sent email to contributors@expensify.com, but still no on my anayetrasel@yahoo.com .

**Comment by johnmlee101** [comment]
@Antasel you'll have to wait for the approval. We typically wait for you to be in these channels before we allow for fully processing payments as a requirement. Note this will not affect any of the payments themselves

**Comment by Antasel** [comment]
@johnmlee101 
I am not worrying about payment, just would like to live in expensify community and get instant help from other devs if needed.
I am getting some new issues on building android/ios

**Comment by sakluger** [comment]
@Antasel got it, thanks for the context. We still need to wait for a slack admin to approve the invite, it's usually done within a day or so.

**Comment by johnmlee101** [comment]
Hi, it should have been approved. Please check your email!

**Comment by Antasel** [comment]
> Hi, it should have been approved. Please check your email!

I sent email again to contributors@expensify.com yesterday. but still no updated. waiting more

**Comment by Antasel** [comment]
@0xmiroslav then no need to make App PR ?

**Comment by Antasel** [comment]
@0xmiroslav Is there any guide to run staging app for desktop, android, iOS ?

**Comment by Antasel** [comment]
thanks.
Where can i download desktop app image ? and for android ?

**Comment by Antasel** [comment]
Ok. I've tested on staging web. working as expected.


[Screencast from 12-9-23 05:25:49+06.webm](https://github.com/Expensify/App/assets/139204153/eb5b8b62-cf7a-4e10-9014-8693ba9755ee)

**Comment by sakluger** [comment]
The [PR](https://github.com/Expensify/expensify-common/pull/573) isn't showing that it was deployed to staging or prod, @0xmiroslav is that specific to Expensfy/expensify-common? I want to make sure the payment date is reflected correctly.

**Comment by Antasel** [comment]
@0xmiroslav thanks for your kind answer

**Comment by sakluger** [comment]
@0xmiroslav thanks!

Copying over the BZ checklist manually since the automation didn't post it in this case:

**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@0xmiroslav] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@0xmiroslav] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@0xmiroslav] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@0xmiroslav] Determine if we should create a regression test for this bug.
- [ ] [@0xmiroslav] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sakluger] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sakluger** [comment]
So excited to close out this issue, thanks everyone!

**Summarizing payouts for this issue:**

Issue reporter: @hungvu193 $250 (hired and paid via Upwork)
Contributor: @Antasel $3000 (hired and paid via Upwork)
Contributor+: @0xmiroslav $3000

Above payments include efficiency bonus 🎉 
Upwork job: https://www.upwork.com/jobs/~013659e025d6ce0e4f

@0xmiroslav - two things:

1. Could you please complete the BZ checklist?
2. Would you like to be paid via NewDot manual request or via Upwork? I see that we sent you an offer [here](https://www.upwork.com/nx/wm/offer/26524537), but it's probably because this issue is so old.

**Comment by sakluger** [comment]
Thanks! I completed the payment through Upwork.
